### ISSUE TITLE
How to do downsampling?

### ISSUE THREAD

**Issue by patrickvonplaten** [HEnquist/rubato#17]
Hello, 

Thanks a lot for this crate!

The provided upsampling example from 44100 -> 48000 works very well, however I was wondering how one can do downsampling. Slighly adapting your code example to downsample from 48000 -> 16000 as follows:

```rust
    let params = InterpolationParameters {
        sinc_len: 256,
        f_cutoff: 0.95,
        interpolation: InterpolationType::Nearest,
        oversampling_factor: 160,
        window: WindowFunction::BlackmanHarris2,
    };
    let mut resampler = SincFixedIn::<f64>::new(
        16000 as f64 / 48000 as f64,
        params,
        1024,
        2,
    );

    let waves_in = vec![vec![0.0f64; 1024];2];
    let waves_out = resampler.process(&waves_in).unwrap();
```

throws this error:

```
thread 'main' panicked at 'Tried to interpolate for index 1281, max for the given input is 1279', /home/patrick/.cargo/registry/src/github.com-1ecc6299db9ec823/rubato-0.8.0/src/interpolator_avx.rs:146:9
```

I would be very grateful if you could provide a downsampling example using your library!

Thanks

**Comment by HEnquist** [comment]
It's supposed to work like you tried. It seems you have discovered a bug! I'll take a look.
Do you need the asynchronous resampler, or could you use the synchronous one (FftFixedIn) instead? That one is faster and probably doesn't have the same bug.

**Comment by patrickvonplaten** [comment]
Hey @HEnquist,

Thanks for the fast reply. I also tried to use `FftFixedIn` instead, but there also seems to be a problem. E.g. the following code produces an empty vector:

```rust
    use rubato::{FftFixedIn, Resampler};

    let samples = vec![vec![0.0f32; 1024]];

    let mut resampler = FftFixedIn::<f32>::new(48000, 16000, 1024, 1, 1);

    let re_samples = resampler.process(&samples).unwrap();

    println!("{:?}", re_samples);  // print vector
```

**Comment by HEnquist** [comment]
Try calling the same resampler over and over in a loop, to give it more data. It will only return something once it has enough samples.

**Comment by HEnquist** [comment]
This PR should fix the issue with the SincFixedIn: https://github.com/HEnquist/rubato/pull/18
Could I ask you to try your downsampling again with the new version in the "issue17" branch?

**Comment by patrickvonplaten** [comment]
Hey, I tried downsampling again with `FftFixedIn`, but it still seems to yield empty vectors. The PR only handles the asynchronus case no? Isn't `FftFixedIn` the synchronus part (no sure though 😅)

**Comment by patrickvonplaten** [comment]
Ah ok sorry, I misunderstood! I'm actually looking for a function that I don't have to call over and over for resampling - it should work with just a single example no?

**Comment by HEnquist** [comment]
Ok! This lib is mainly intended for resampling of continuous streams. So you have a bunch of shorter clips that you want to resample, right? You can use SincFixedIn for that, but there are a few things to keep in mind.
Let's say we have a 3000 sample clip that we want to resample from 48kHz to 16 kHz. We create a SincFixedIn with chunksize 3000, and sinc length 128. When resampling that we would like 1000 samples in the result, but because of the interpolation filter (the sinc) it will have to stop at 128/2 from the end of the input. This means the length of the resampled data will be (3000-64)/2, about 980. The remaining samples are stored for the next call, and in the next call it will be able to use the stored samples to give you 1000 samples.
The easuest way to resample the 3000 sample clip is to do as above, and call the resampler once. Then call it again with 3000 zeros. Take the first 20-ish samples and append them to the first result, and trash the rest.

**Comment by HEnquist** [comment]
Closing as the panic is solved in 0.8.1
